src/docbuilder/caching.rs

```diff
@@ -0,0 +1,112 @@
+use crate::utils::copy_dir_all;
+use anyhow::{Context as _, Result};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+};
+use tracing::{debug, instrument, warn};
+
+/// move cache folder to target, falling back to copy + delete on error.
+fn move_or_copy<P: AsRef<Path> + std::fmt::Debug, Q: AsRef<Path> + std::fmt::Debug>(
+    source: P,
+    dest: Q,
+) -> io::Result<()> {
+    if let Some(parent) = dest.as_ref().parent() {
+        fs::create_dir_all(parent)?;
+    }
+    if let Err(err) = fs::rename(&source, &dest) {
+        warn!(
+            ?err,
+            ?source,
+            ?dest,
+            "could not move target directory, fall back to copy"
+        );
+        copy_dir_all(&source, &dest)?;
```

Comment:
if we can't `symlink` or `move`, we probably should compress the target directory before copying it.